### PR TITLE
API to dynamically manage timeline markers

### PR DESCRIPTION
**Pull request by protyposis**
Implements #103

### CODE REVIEW

File: src/ts/components/seekbar.ts [hunk 1]
```diff
@@ -416,24 +422,18 @@ export class SeekBar extends Component<SeekBarConfig> {
...
-      if (markers && player.getDuration() !== Infinity) {
-        for (let marker of markers) {
-          this.timelineMarkers.push({
-            time: 100 / player.getDuration() * marker.time, // convert time to percentage
-            title: marker.title,
-          });
-        }
+      const duration = player.getDuration();
+
+      if (duration === Infinity) {
```
Comment (bitmovin-kenny) [inline src/ts/components/seekbar.ts]: I think live videos have `duration == infinity` too, this might break this functionality

Comment (bitmovin-kenny) [reply src/ts/components/seekbar.ts]: I just read in the comment section of the isse that `We don't support markers on live streams right now` ... so i guess this is ok

File: src/ts/components/seekbar.ts [hunk 2]
```diff
@@ -444,6 +444,11 @@ export class SeekBar extends Component<SeekBarConfig> {
     player.addEventHandler(player.EVENT.ON_READY, setupMarkers);
     // Remove markers when unloaded
     player.addEventHandler(player.EVENT.ON_SOURCE_UNLOADED, clearMarkers);
+    // Remove markers when unloaded
```
Comment (bitmovin-kenny) [inline src/ts/components/seekbar.ts]: duplicate comment (does not match the `onResize` subscription)

File: src/ts/components/seekbar.ts [hunk 3]
```diff
@@ -607,30 +612,49 @@ export class SeekBar extends Component<SeekBarConfig> {
...
+        'data-marker-time': String(marker.marker.time),
+        'data-marker-title': String(marker.marker.title),
+      }).css(cssProperties));
     }
   }
 
-  protected getMarkerAtPosition(percentage: number): TimelineMarker | null {
-    let snappedMarker: TimelineMarker = null;
-    let snappingRange = 1;
+  protected getMarkerAtPosition(percentage: number): SeekBarMarker | null {
```
Comment (bitmovin-kenny) [inline src/ts/components/seekbar.ts]: should it be possible to have overlapping markers? if so we must not break in the for loop bur rather collect all matching markers

File: src/ts/uimanager.ts [hunk 4]
```diff
@@ -461,6 +540,35 @@ export class UIManager {
...
+   */
+  getTimelineMarkers(): TimelineMarker[] {
+    return this.config.metadata.markers;
+  }
+
+  /**
+   * Adds a marker to the timeline. Does not check for duplicates/overlaps at the `time`.
+   */
+  addTimelineMarker(timelineMarker: TimelineMarker): void {
+    this.config.metadata.markers.push(timelineMarker);
```
Comment (bitmovin-kenny) [inline src/ts/uimanager.ts]: if we do not allow overlapping markers, we need to check the ranges here

File: src/ts/components/seekbar.ts [hunk 5]
```diff
@@ -607,30 +612,49 @@ export class SeekBar extends Component<SeekBarConfig> {
...
-          snappedMarker = marker;
-          break;
+        if (marker.duration > 0
+          && percentage >= marker.position - snappingRange
+          && percentage <= marker.position + marker.duration + snappingRange) {
+          return marker;
+        }
+        else if (percentage >= marker.position - snappingRange
+          && percentage <= marker.position + snappingRange) {
+          return marker;
```
Comment (Reevn) [inline src/ts/components/seekbar.ts]: I'm not a fan of multi-line if statements. How about something more expressive and clean?
```ts
if (this.timelineMarkers.length > 0) {
  for (let marker of this.timelineMarkers) {
    const isMarkerInStartRange = percentage >= marker.position - snappingRange;
    const isMarkerInEndRange = percentage <= marker.position + marker.duration + snappingRange;

    if (marker.duration > 0 && isMarkerInStartRange && isMarkerInEndRange) {
      return marker;
    } else if (isMarkerInStartRange && percentage <= marker.position + snappingRange) {
      return marker;
    }
  }
}
```

The variable names could of course be chosen better etc. And the example could be written better but you get my point 👍

File: src/ts/uimanager.ts [hunk 6]
```diff
@@ -461,6 +540,35 @@ export class UIManager {
...
+
+  /**
+   * Returns the list of all added markers in undefined order.
+   */
+  getTimelineMarkers(): TimelineMarker[] {
+    return this.config.metadata.markers;
+  }
+
+  /**
+   * Adds a marker to the timeline. Does not check for duplicates/overlaps at the `time`.
```
Comment (Reevn) [inline src/ts/uimanager.ts]: Is there a use-case for not filtering exact duplicates before adding markers?

File: src/ts/components/seekbar.ts (hunk 1 above)
Comment (protyposis) [reply src/ts/components/seekbar.ts]: This infinity check was there before, I just rearranged the code. Never seemed to be an issue, but you're right regarding live streams. Since we don't support them yet, I'd say we leave that for another PR.

File: src/ts/components/seekbar.ts (hunk 3 above)
Comment (protyposis) [reply src/ts/components/seekbar.ts]: It is possible (overlaps are not prevented), but it does not make sense and I leave the responsibility to manage that correctly to the user. Also, which `SeekBarLabel` for which marker should be displayed then? All of them? Let's also leave this open until it becomes an issue.

File: src/ts/uimanager.ts (hunk 4 above)
Comment (protyposis) [reply src/ts/uimanager.ts]: As explained above.

File: src/ts/components/seekbar.ts (hunk 5 above)
Comment (protyposis) [reply src/ts/components/seekbar.ts]: I wrote it on purpose the way it is because in your proposal the `isMarkerInEndRange` is always calculated although only needed in case `marker.duration > 0` 😝 But you're right, it's hard to follow.

File: src/ts/uimanager.ts (hunk 6 above)
Comment (protyposis) [reply src/ts/uimanager.ts]: Yes, I leave that responsibility to the user. If he really wants to have overlaps he can have them. But it does not make sense and will probably never happen in practice, because chapters or ad breaks seldomly overlap. That is unless somebody figures out that these markers can be used for some other creative stuff I didn't think of, but then we can always improve later.
